### PR TITLE
feat: Add Deno runtime adapter with cross-runtime CI and JSR publishing

### PR DESCRIPTION
**Pull request by kjanat**
## Overview

Adds first-class Deno support, TypeScript-first module resolution, cross-runtime CI, and JSR publishing. Wires a full Deno runtime adapter into runtime discovery, updates imports/tests to use .ts sources, and adds workflows/scripts to validate and publish a Deno-first package alongside existing npm workflows.

## Key Changes

- Deno runtime adapter
  - New implementation: src/runtime/deno.ts (exports type DenoNamespace and factory createDenoAdapter(ns?))
  - Permission-safe helpers (safeEnvToObject, safeCwd), synthetic argv (['deno','run', ...]), readFile semantics (null for NotFound/PermissionDenied), exit delegation, stdin line-reading, stdout/stderr wrappers
  - Minimal ambient typings: src/runtime/deno-builtins.d.ts
  - Extensive unit tests: src/runtime/deno.test.ts (many edge cases: perms, I/O, homedir/configDir)
  - Test helper for mocking Deno global: src/runtime/test-helpers.ts (withMockDenoGlobal)
  - Auto-detect now returns a Deno adapter when appropriate: src/runtime/auto.ts
  - Runtime barrels re-export Deno symbols: src/runtime/index.ts, src/runtime.ts

- Cross-runtime CI & tests
  - CI workflow: .github/workflows/ci.yml — concurrency, Bun-based lint/typecheck/format, Node/Bun test matrix, Deno smoke test against built output, build step
  - Deno smoke test script: scripts/deno-smoke-test.ts (runs under Deno against dist/runtime.mjs)
  - Tests and imports updated to .ts module specifiers and adjusted for Deno mocking

- JSR publishing
  - Publish workflow: .github/workflows/publish-jsr.yml (release/manual, sets up Deno/Bun, builds, runs smoke test, deno publish with OIDC)
  - deno.json added with package metadata, subpath exports, publish include/exclude, and nodeModulesDir/manual guidance

- Project & type improvements
  - package.json version bumped to 0.9.0
  - tsconfig.json: noEmit and allowImportingTsExtensions enabled (emit-related options removed)
  - Explicit type annotations hardened public constants (RUNTIMES, SHELLS) to satisfy JSR/type-check constraints
  - Widespread module specifier changes from .js → .ts across source and tests to support TypeScript-first resolution

- Docs & project state
  - AGENTS.md / CHANGELOG.md updated
  - .opencode state files added documenting the v0.9 plan and tasks

## Public API Surface Changes

- New public exports:
  - createDenoAdapter(ns?: DenoNamespace)
  - type DenoNamespace
- Existing public exports preserved; some exported constant types made more explicit (RUNTIMES, SHELLS). No other breaking changes observed.

## Testing & Validation

- Deno adapter covered by unit tests (vitest with mock Deno namespace) and a Deno-native smoke test that runs against built output in CI/publish workflows.
- CI runs lint/typecheck/format under Bun, vitest across Node and Bun, and a Deno smoke test to ensure cross-runtime parity.

## Notable Added Files

- .github/workflows/ci.yml
- .github/workflows/publish-jsr.yml
- deno.json
- scripts/deno-smoke-test.ts
- src/runtime/deno.ts
- src/runtime/deno-builtins.d.ts
- src/runtime/deno.test.ts
- src/runtime/test-helpers.ts

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,51 @@ adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
...
+#### Cross-Runtime CI
+
+- **GitHub Actions CI workflow** (`.github/workflows/ci.yml`) — lint+typecheck (Bun), test matrix
+  (Node LTS + Bun stable), Deno smoke test, build with publint+attw.
+- **Deno smoke test** (`scripts/deno-smoke-test.ts`) — runs on real Deno runtime after build,
+  exercising `createDenoAdapter()` against actual Deno APIs.
+
+#### JSR Publishing
+
+- **`deno.json`** with JSR package config (`@dreamcli/dreamcli`), three subpath exports, publish
```
[the remark this answers is not in the document]
Comment (kjanat) [reply CHANGELOG.md]: Good find, I want it to be `@kjanat/dreamcli`

File: deno.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+	"name": "@kjanat/dreamcli",
+	"version": "0.9.0",
+	"exports": {
+		".": "./src/index.ts",
+		"./runtime": "./src/runtime.ts",
+		"./testkit": "./src/testkit.ts"
+	},
+	"publish": {
+		"include": ["src", "LICENSE"],
+		"exclude": ["src/**/*.test.ts", "src/**/AGENTS.md"]
```
[the remark this answers is not in the document]
Comment (kjanat) [reply deno.json]: There is no README.md yet

File: src/core/cli/cli-completion-e2e.test.ts [hunk 3]
```diff
@@ -6,15 +6,14 @@
...
-import { middleware } from '../schema/middleware.js';
-import { cli } from './index.js';
+import { describe, expect, it, vi } from 'vitest';
+import { createTestAdapter, ExitError } from '../../runtime/adapter.ts';
+import type { GlobalForDetect } from '../../runtime/detect.ts';
+import { arg } from '../schema/arg.ts';
+import { command } from '../schema/command.ts';
+import { flag } from '../schema/flag.ts';
+import { middleware } from '../schema/middleware.ts';
+import { cli } from './index.ts';
```
[the remark this answers is not in the document]
Comment (kjanat) [reply src/core/cli/cli-completion-e2e.test.ts:16]: Adjust your guidelines. I want `.ts`.

File: src/core/completion/index.ts [hunk 4]
```diff
@@ -36,10 +36,12 @@ type Shell = 'bash' | 'zsh' | 'fish' | 'powershell';
...
-const SHELLS = Object.freeze(['bash', 'zsh', 'fish', 'powershell'] as const satisfies readonly [
-	Shell,
-	...Shell[],
-]);
+const SHELLS: Readonly<readonly ['bash', 'zsh', 'fish', 'powershell']> = Object.freeze([
+	'bash',
+	'zsh',
+	'fish',
+	'powershell',
+] as const satisfies readonly [Shell, ...Shell[]]);
```
[the remark this answers is not in the document]
Comment (kjanat) [reply src/core/completion/index.ts:44]: Well... Stop twitching... I ran the formatter, and this came out, so I'm fine with it.

File: src/testkit.ts [hunk 5]
```diff
@@ -7,11 +7,11 @@
...
-export type { TestAdapterOptions } from './runtime/index.js';
-export { createTestAdapter } from './runtime/index.js';
+export type { CapturedOutput } from './core/output/index.ts';
+export { createCaptureOutput } from './core/output/index.ts';
+export type { TestAnswer, TestPrompterOptions } from './core/prompt/index.ts';
+export { createTestPrompter, PROMPT_CANCEL } from './core/prompt/index.ts';
+export type { RunOptions, RunResult } from './core/testkit/index.ts';
+export { runCommand } from './core/testkit/index.ts';
+export type { TestAdapterOptions } from './runtime/index.ts';
+export { createTestAdapter } from './runtime/index.ts';
```
[the remark this answers is not in the document]
Comment (kjanat) [reply src/testkit.ts:17]: Two parts to the suggestion, two problems:

1. "Stop exporting from the runtime index" — The reviewer assumes src/runtime/index.ts is a public surface. It's not. The public subpath is src/runtime.ts, which already excludes createTestAdapter/TestAdapterOptions. The internal barrel (src/runtime/index.ts) exports them so that src/testkit.ts can re-import them — that's the barrel doing its job.
 2. "Import from adapter module directly instead of barrel" — This is the project's circular-dep-avoidance convention being over-applied. The convention exists to prevent cycles. Testkit is a leaf node in the dependency graph — no cycle risk. Barrel import is the normal, conventional path when cycles aren't a concern.

The review's mental model was: "runtime index exports test symbols → that's wrong, remove them." But the layering is actually:

```
     src/runtime/adapter.ts    ← defines createTestAdapter
            ↓
     src/runtime/index.ts      ← internal barrel, re-exports everything
            ↓
     src/testkit.ts            ← public subpath, cherry-picks test symbols
     src/runtime.ts            ← public subpath, excludes test symbols ✓
```

The public API is already correct. The suggested change would just swap the import path in testkit.ts from ./runtime/index.ts to ./runtime/adapter.ts — identical result, zero functional difference.

File: src/runtime/test-helpers.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+			readable: {
+				getReader: () => ({
+					read: () => Promise.resolve({ done: true, value: undefined }),
+					releaseLock: () => {},
+				}),
+			},
+		},
+		exit: vi.fn() as unknown as (code: number) => never,
+		readTextFile: () => Promise.reject(Object.assign(new Error('not found'), { name: 'NotFound' })),
+		version: { deno: '2.0.0' },
```
[the remark this answers is not in the document]
Comment (kjanat) [reply src/runtime/test-helpers.ts:37]: The diagnosis was half-right, the prescription was wrong.

What the reviewer got right: test-helpers.ts doesn't match the publish exclude pattern `src/**/*.test.ts`, so it would ship — and a published file with `import { vi } from 'vitest'` breaks at runtime for consumers. That's a real problem.

What the reviewer got wrong: The conclusion. The file is test-only infrastructure (`@internal` — test-only, not exported from any public barrel).
The fix isn't to rewrite it without vitest — it's to exclude it from publishing. We added "src/**/test-helpers.ts" to deno.json's `publish.exclude` (1 line).

The reviewer's suggestion — hand-rolling spy factories to replace vi.fn() — would:

- Reinvent what vitest already provides
- Make the test helper harder to maintain
- Lose vi.fn() features (call tracking, toHaveBeenCalledWith, etc.)
- Solve a publishing problem with a test-code rewrite instead of a config fix
- The zero-dep constraint applies to production code. Test helpers importing test frameworks is normal — they just shouldn't be published.

File: src/runtime/index.ts [hunk 7]
```diff
@@ -9,11 +9,13 @@
...
-export { createTestAdapter, ExitError } from './adapter.js';
-export { createAdapter } from './auto.js';
-export { createBunAdapter } from './bun.js';
-export type { GlobalForDetect, Runtime } from './detect.js';
-export { detectRuntime, RUNTIMES } from './detect.js';
-export type { NodeProcess } from './node.js';
-export { createNodeAdapter } from './node.js';
+export type { RuntimeAdapter, TestAdapterOptions } from './adapter.ts';
+export { createTestAdapter, ExitError } from './adapter.ts';
+export { createAdapter } from './auto.ts';
```
[the remark this answers is not in the document]
Comment (kjanat) [reply src/runtime/index.ts:14]: Two parts to the suggestion, two problems:

1. "Stop exporting from the runtime index" — The reviewer assumes `src/runtime/index.ts` is a public surface. It's not. The public subpath is `src/runtime.ts`, which already excludes `createTestAdapter`/`TestAdapterOptions`. The internal barrel (`src/runtime/index.ts`) exports them so that `src/testkit.ts` can re-import them — that's the barrel doing its job.
2. "Import from adapter module directly instead of barrel" — This is the project's circular-dep-avoidance convention being over-applied. The convention exists to prevent cycles. Testkit is a leaf node in the dependency graph — no cycle risk. Barrel import is the normal, conventional path when cycles aren't a concern.

The review's mental model was: "runtime index exports test symbols → that's wrong, remove them." But the layering is actually:

```
src/runtime/adapter.ts    ← defines createTestAdapter
       ↓
src/runtime/index.ts      ← internal barrel, re-exports everything
       ↓
src/testkit.ts            ← public subpath, cherry-picks test symbols
src/runtime.ts            ← public subpath, excludes test symbols ✓
```

The public API is already correct. The suggested change would just swap the import path in `testkit.ts` from `./runtime/index.ts` to `./runtime/adapter.ts` — identical result, zero functional difference.

File: scripts/deno-smoke-test.ts [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+assert(typeof adapter.readFile === 'function', 'readFile is a function');
+assert(typeof adapter.homedir === 'string', 'homedir is a string');
+assert(typeof adapter.configDir === 'string', 'configDir is a string');
+
+// readFile should return contents for existing file
+const pkg = await adapter.readFile('./package.json');
+assert(pkg !== null, 'readFile returns content for existing file');
+assert(typeof pkg === 'string' && pkg.includes('dreamcli'), 'readFile content contains "dreamcli"');
+
```
[the remark this answers is not in the document]
Comment (kjanat) [reply scripts/deno-smoke-test.ts:108]: Count again... line length is 100 exactly...

File: scripts/deno-smoke-test.ts [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+ * Usage: deno run --allow-read --allow-env scripts/deno-smoke-test.ts
+ *
+ * Exit codes:
+ *   0 — all checks passed
+ *   1 — a check failed
+ */
+
+// Import from built output (not source) to verify the published shape
+import type { RuntimeAdapter } from '../dist/runtime.mjs';
+import { createDenoAdapter } from '../dist/runtime.mjs';
```
[the remark this answers is not in the document]
Comment (kjanat) [reply scripts/deno-smoke-test.ts:17]: Dude... the build output doesn't output `.js`, so tell me how to import it...
